lib/Sema/CSGen.cpp

```diff
@@ -1345,8 +1345,10 @@ namespace {
       // We may, alternatively, want to use a type variable in that case,
       // and possibly infer the type of the variable that way.
       CS.getTypeChecker().validateDecl(E->getDecl());
-      if (E->getDecl()->isInvalid())
+      if (E->getDecl()->isInvalid()) {
+        E->setType(E->getDecl()->getInterfaceType());
         return nullptr;
+      }
```

Comment:
I've been working on moving the constraint system over to use a side map for types, which are accessed via ConstraintSystem::getType(Expr *) and set via ConstraintSystem::setType(Expr *, Type), so those should be used here (at the moment they still set the type directly on the expression, but that will be going away).